payments.md

```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
+    ToSend TokenAmount
+
+    ClosingAt uint64
+    MinCloseHeight uint64
+
+    LaneStates map[uint64]LaneState
+}
+
+type LaneState struct {
+    Nonce uint64
+    Redeemed TokenAmount
+}
+
+func (paych *PaymentChannel) validateSignature(sv SpendVoucher) {
+    if msg.From == paych.From {
+        ValidateSignature(sv.SerializeNoSig(), sv.Signature, paych.To)
+    } else if msg.From == paych.To {
+        ValidateSignature(sv.SerializeNoSig(), sv.Signature, paych.From)
+    } else {
+        Fatal("bad programmer")
+    }
+}
+
+func (paych *PaymentChannel) UpdateChannelState(sv SpendVoucher, secret []byte, pip *PieceInclusionProof) {
+    if !paych.validateSignature(sv) {
+        Fatal("Signature Invalid")
+    }
+
+    if chain.Now() < sv.TimeLock {
+        Fatal("cannot use this voucher yet!")
+    }
+
+    if sv.SecretPreimage != nil {
+        if Hash(secret) != sv.SecretPreimage {
+            Fatal("Incorrect secret!")
+        }
+    }
+  
+  	if sv.DataCommitment != nil {
+      // Checks that the piece inclusion proof is valid, and that the referenced sector
+      // is correctly being stored
+      if !ValidateInclusion(pip, sv.DataCommitment) {
+        Fatal("PieceInclusionProof was invalid")
+      }
+  	}
+
+    if sv.RequiredSector != nil {
+        miner, found := GetMiner(msg.From)
+        if !found {
+            Fatal("Redeemer is not a miner")
+        }
+
+        if !miner.HasSector(sv.RequiredSector) {
+            Fatal("miner does not have sector, cannot redeem payment")
+        }
+    }
+
+    ls := paych.LaneStates[sv.Lane]
+    if ls.Closed {
+        Fatal("cannot redeem a voucher on a closed lane")
+    }
+
+    if ls.Nonce > sv.Nonce {
+        Fatal("voucher has an outdated nonce, cannot redeem")
+    }
+
+    var mergeValue TokenAmount
+    for _, merge := range sv.Merges {
+        ols := paych.LaneStates[merge.Lane]
+
+        if ols.Nonce >= merge.Nonce {
+            Fatal("merge in voucher has outdated nonce, cannot redeem")
+        }
+
+        mergeValue += ols.Redeemed
+        ols.Nonce = merge.Nonce
+    }
+
+    ls.Nonce = sv.Nonce
+    balanceDelta = sv.Amount - (mergeValue + ls.Redeemed)
+    ls.Redeemed = sv.Amount
+
+    newSendBalance = paych.ToSend + balanceDelta
+    if newSendBalance < 0 {
+        // TODO: is this impossible?
+        Fatal("voucher would leave channel balance negative")
+    }
+
+    if newSendBalance > paych.ChannelTotal {
+        Fatal("not enough funds in channel to cover voucher")
+    }
+
+    paych.ToSend = newSendBalance
+
+    if sv.MinCloseHeight != 0 {
+        if paych.ClosingAt < sv.MinCloseHeight {
+            paych.ClosingAt = sv.MinCloseHeight
+        }
+        if paych.MinCloseHeight < sv.MinCloseHeight {
+            paych.MinCloseHeight = sv.MinCloseHeight
+        }
+    }
+}
+
+func (paych *PaymentChannel) Withdraw(upTo TokenAmount) {
+    // TODO: this ones tricky, withdraw funds without closing it out entirely...
+}
```

Comment:
is this the same as collect?

Reply:
probably going to delete this one anyways. its difficult to do right, and we can just close channels out